File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0,r = heights.size()-1;
        int ans=0,tmp;
        while(l<r){
            tmp = (r-l)*min(heights[l],heights[r]);
            if(heights[l]<heights[r])l++;
            else r--;
            ans=max(ans,tmp);
        }
        return ans;

    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        ans =0
        n = len(heights)
        stack = []
        for i in range(n):
            start = i
            while stack and stack[-1][1] > heights[i]:
                ind = stack[-1][0]
                height = stack[-1][1]
                stack.pop()
                ans = max(ans,height*(i-ind))
                start = ind
            stack.append((start,heights[i]))
        
        for i in range(len(stack)):
            h = stack[i][1]
            ind = stack[i][0]
            ans = max(ans,h*(n-ind))
        
        return ans